util/compression.h

```diff
@@ -216,36 +217,53 @@ struct CompressionDict {
 
 // Holds dictionary and related data, like ZSTD's digested uncompression
 // dictionary.
-struct UncompressionDict {
+struct UncompressionDict : public Cleanable {
+  // Block containing the data for the compression dictionary. It is non-empty
+  // only if the constructor that takes a string parameter is used.
+  std::string dict_;
```

Comment:
Trying to understand the change here.
If dict_ is empty, it means the data of the dictionary is in block cache. It also implicitly indicates that UncompressionDict does not own it.
If dict_ is not empty, it means UncompressionDict owns the data, and it means the data is not managed in the block cache?
Am I missing anything?

Reply:
There are two cases:

1) The string constructor is used (as in BlockBasedTableBuilder): dict_ contains the data, slice_ points to dict_, and UncompressionDict is self-contained.
2) The Slice constructor is used (as on the read code paths): in this case dict_ is empty, slice_ is whatever was passed in, and UncompressionDict is not necessarily self-contained. However, since UncompressionDict is now a Cleanable, on the read paths we can and do transfer the ownership of the actual compression dictionary bytes from the CachableEntry to the UncompressionDict itself, once again creating a self-contained UncompressionDict object.

I will extend the comments to make this clearer.

Reply:
P.S. In the case of the block cache, by "management" I mean releasing the cache handle when the UncompressionDict is destroyed.

Reply:
Added comments to the UncompressionDict constructors to explain the semantics.